pkg/provider/onepassword/onepassword.go

```diff
@@ -0,0 +1,364 @@
+	errExpectedOneItem                    = "expected one 1Password Item matching %w"
+	errGetItem                            = "error finding 1Password Item: %w"
+	errKeyNotFound                        = "key not found in 1Password Vaults: %w"
+	errDocumentNotFound                   = "error finding 1Password Document: %w"
+	errExpectedOneField                   = "expected one 1Password ItemField matching %w"
+
+	documentCategory = "DOCUMENT"
+)
+
+// ProviderOnePassword is a provider for 1Password.
+type ProviderOnePassword struct {
+	vaults map[string]int
+	client connect.Client
+}
+
+// https://github.com/external-secrets/external-secrets/issues/644
+var _ esv1beta1.SecretsClient = &ProviderOnePassword{}
+var _ esv1beta1.Provider = &ProviderOnePassword{}
+
+// NewClient constructs a 1Password Provider.
+func (provider *ProviderOnePassword) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	config := store.GetSpec().Provider.OnePassword
+
+	credentialsSecret := &corev1.Secret{}
+	objectKey := types.NamespacedName{
+		Name:      config.Auth.SecretRef.ConnectToken.Name,
+		Namespace: namespace,
+	}
+
+	// only ClusterSecretStore is allowed to set namespace (and then it's required)
+	if store.GetObjectKind().GroupVersionKind().Kind == esv1beta1.ClusterSecretStoreKind {
+		if config.Auth.SecretRef.ConnectToken.Namespace == nil {
+			return nil, fmt.Errorf(errClusterSecretStoreNamespaceMissing)
+		}
+		objectKey.Namespace = *config.Auth.SecretRef.ConnectToken.Namespace
+	}
+
+	err := kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return nil, fmt.Errorf(errFetchK8sSecret, err)
+	}
+	token := credentialsSecret.Data[config.Auth.SecretRef.ConnectToken.Key]
+	if (token == nil) || (len(token) == 0) {
+		return nil, fmt.Errorf(errMissingToken)
+	}
+	provider.client = connect.NewClientWithUserAgent(config.ConnectHost, string(token), userAgent)
+	provider.vaults = config.Vaults
+
+	return provider, nil
+}
+
+// ValidateStore checks if the provided store is valid.
+func (provider *ProviderOnePassword) ValidateStore(store esv1beta1.GenericStore) error {
+	return validateStore(store)
+}
+
+func validateStore(store esv1beta1.GenericStore) error {
+	storeSpec := store.GetSpec()
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.OnePassword == nil {
+		return fmt.Errorf(errOnePasswordStore)
+	}
+	config := storeSpec.Provider.OnePassword
+
+	// only ClusterSecretStore is allowed to set namespace (and then it's required)
+	if store.GetObjectKind().GroupVersionKind().Kind == esv1beta1.ClusterSecretStoreKind {
+		if config.Auth.SecretRef.ConnectToken.Namespace == nil {
+			return fmt.Errorf(errClusterSecretStoreNamespaceMissing)
+		}
+	} else {
+		if config.Auth.SecretRef.ConnectToken.Namespace != nil {
+			return fmt.Errorf(errSecretStoreNamespaceSet)
+		}
+	}
+
```

Comment:
We can add provider specific checks here. We can check that the connect host string is a valid URL with url.ParseRequestURI, also check if mandatory fields are passed, check if there is at least one vault passed, and so on.

Reply:
Do I understand correctly that we already get validation when not using `omitEmpty`? Almost all the fields are required.

I did setup tests and checks for `ConnectToken.Name` and `ConnectToken.Key` and Parse url, good call.

Reply:
Sorry, I was just listing things that could be generally added to this method based on what we did with other providers. Checking if fields are not nil was done mostly when we have fields that are optional with the kubebuilder flags, but at least one of them have to be passed in the end for the provider to work. This is not 1pass case.

Reply:
There is one caveat with checking if fields are not nil. Kubebuider will allow required string fields to be empty strings, such as this example below:
```
apiVersion: external-secrets.io/v1beta1
kind: SecretStore
metadata:
  name: example-auth
spec:
  provider:
    oracle:
      vault: ""
      region: "bar"
      auth:
        user: "user"
        tenancy: "ten"
        secretRef:
          privatekey:
            name: "priv"
            key: "key"
          fingerprint:
            name: "fing"
            key: "key"                                                                                                                                                               
```
Oracle's `NewClient` implementation in this case does a sanity check verifying if the field is not empty. This is also something we are considering to put in the `ValidateStore` method at #1083, as a way to catch the config error faster. 

I guess we could configure kubebuilder flags with `+kubebuilder:validation:MinLength=1` as well, though the error message would be a little bit less obvious.

Not sure if this happens with 1password provider, but the other thing we might need to check at `ValidateStore` is multiple configuration setups (if they are allowed). Some validation might need to be happening just to be sure users are properly configuring things.

Reply:
For a valid 1Password SecretStore/ClusterSecretStore, the only field that is not required all of the time is `connectTokenSecretRef.namespace` which we're now checking with `utils.ValidateSecretSelector`.

I believe I'm now checking every other field as well. `connectHost` and `vaults` are being checked. I think that's it, correct me if I'm missing something. See `apis/externalsecrets/v1beta1/secretstore_onepassword_types.go` in this PR.

> the other thing we might need to check at ValidateStore is multiple configuration setups (if they are allowed).

@gusfcarvalho Sorry, I don't follow. Multiple configuration setups?

Reply:
The only variation on SecretStore configuration is how many vaults you put in the list. - That does make me realize though that I should add a check for non-unique numbers in that list. Doing that now.

Reply:
OK. I just added a `hasUniqueVaultNumbers` check and related tests.

There's no way to check if they chose the right connectHost, but there's stuff about this in the docs.